src/test/groovy/com/ullink/NuGetDownloadTest.groovy

```diff
@@ -0,0 +1,217 @@
+    }
+
+    @After
+    void cleanup() {
+        if (tempFolder != null && tempFolder.exists()) {
+            tempFolder.deleteDir()
+        }
+    }
+
+    @Test
+    void shouldBeAbleToDownloadNugetWithoutAnyParameters() {
+        // When
+        executeSomeNugetTask()
+
+        // Then
+        Assert.assertTrue(nugetExecutableExistsInCache())
+    }
+
+    @Test
+    void shouldThrowErrorOnInvalidHttpPath() {
+        // When
+        project.nugetPack {
+            nugetExePath = "http://localhost:99999"
+        }
+
+        // Then
+        taskShouldBeExecutedWithException()
+
+        Assert.assertFalse(nugetExecutableExistsInCache())
+    }
+
+    /**
+     * Idea of this test:
+     * 1. Download nuget.exe to the some location. This simulates situation, when nuget.exe was already installed on CI
+     * 2. Ask plugin just to use existing nuget.exe file from the predefined location
+     * 3. Verify that:
+     *    * Task was executed without exceptions (e.g. some nuget file was used)
+     *    * Nuget caches folder does not have nuget file (e.g. task used some other nuget file)
+     */
+    @Test
+    void shouldBeAbleToCopyPreviouslyDownloadedNugetFile() {
+        // Given
+        downloadNugetExeToTemp("https://dist.nuget.org/win-x86-commandline/v3.3.0/nuget.exe")
+
+        // When
+        project.nugetPack {
+            nugetExePath = simulatedNugetExecutablePath.absolutePath
+        }
+
+        // Then
+        executeSomeNugetTask()
+
+        def cachedNugetFile = getNugetExeIfExist()
+
+        Assert.assertNull(cachedNugetFile)
+    }
+
+    /**
+     * Idea of this test:
+     * 1. Ask plugin to download nuget.exe from the custom location
+     * 2. Download the same file manually
+     * 3. Compare checksums of files
+     */
+    @Test
+    void shouldDownloadNugetFromCustomUrl() {
+        // Given
+        def customUrl = "https://dist.nuget.org/win-x86-commandline/v3.3.0/nuget.exe"
+
+        // When
+        project.nugetPack {
+            nugetExePath = customUrl
+        }
+
+        // Then
+        executeSomeNugetTask()
+
+        def cachedNugetFile = getNugetExeIfExist()
+
+        Assert.assertNotNull(cachedNugetFile)
+
+        downloadNugetExeToTemp("https://dist.nuget.org/win-x86-commandline/v3.3.0/nuget.exe")
+
+        assertFileAreIdential(cachedNugetFile, simulatedNugetExecutablePath)
+    }
+
+    private static void assertFileAreIdential(File expected, File actual){
+
+        def expectedChecksum = FileUtils.checksum(expected, new CRC32()).value
+        def actualChecksum = FileUtils.checksum(actual, new CRC32()).value
+
+        Assert.assertEquals(expectedChecksum, actualChecksum)
+    }
+
+    private void downloadNugetExeToTemp(String remoteUrl){
+
+        def nonStandardNugetUrl = new URL(remoteUrl)
+
+        FileUtils.copyURLToFile(nonStandardNugetUrl, simulatedNugetExecutablePath)
+
+    }
+
+    private void taskShouldBeExecutedWithException() {
```

Comment:
We could use http://mrhaki.blogspot.ro/2009/11/groovy-goodness-testing-for-expected.html instead of writing our own

Reply:
done